core/templates/components/checkpoint-celebration-modal/checkpoint-celebration-modal.component.ts

```diff
@@ -0,0 +1,312 @@
+    this.generateCheckpointStatusArray();
+
+    this.checkpointCelebrationUtilityService.setIsOnCheckpointedState(true);
+    if (this.shouldDisplayFullScaleMessage) {
+      this.triggerStandardMessage();
+    } else {
+      this.triggerMiniMessage();
+    }
+  }
+
+  generateCheckpointStatusArray(): void {
+    for (let i = 0; i < this.currentCheckpointPosition; i++) {
+      this.checkpointStatusArray[i] = CHECKPOINT_STATUS_COMPLETED;
+    }
+    if (this.totalNumberOfCheckpoints > this.currentCheckpointPosition) {
+      this.checkpointStatusArray[this.currentCheckpointPosition] = (
+        CHECKPOINT_STATUS_IN_PROGRESS);
+    }
+    for (
+      let i = this.currentCheckpointPosition + 1;
+      i < this.totalNumberOfCheckpoints;
+      i++
+    ) {
+      this.checkpointStatusArray[i] = CHECKPOINT_STATUS_INCOMPLETE;
+    }
+    this.checkpointNodesAreVisible = true;
+  }
+
+  getCompletedProgressBarWidth(): number {
+    if (!this.messageModalIsShown || this.messageModalIsDismissed) {
+      return 0;
+    }
+    if (this.currentCheckpointPosition === 0) {
+      return 0;
+    }
+    const spaceBetweenEachNode = 100 / (this.totalNumberOfCheckpoints - 1);
+    return (
+      ((this.currentCheckpointPosition - 1) * spaceBetweenEachNode) +
+      (spaceBetweenEachNode / 2));
+  }
+
+  setFadeInDelaysForCheckpointNodes(): void {
+    for (let i = 0; i < this.totalNumberOfCheckpoints; i++) {
+      this.checkpointNodeFadeInDelays[i] = (
+        i * (1.5 / this.totalNumberOfCheckpoints) + 2.2);
+    }
+  }
+
+  triggerStandardMessage(): void {
+    this.resetTimer();
+    this.messageModalIsShown = true;
+    this.autoMessageDismissalTimeout = setTimeout(() => {
+      this.messageModalIsShown = false;
+      this.checkpointNodesAreVisible = false;
+    }, MESSAGE_MODAL_APPROX_COMPLETE_ANIMATION_DURATION_MS);
+  }
+
+  triggerMiniMessage(): void {
+    this.miniMessageTooltipIsShown = true;
+    setTimeout(() => {
+      this.miniMessageTooltipIsShown = false;
+    }, MINI_MESSAGE_MODAL_APPROX_COMPLETE_ANIMATION_DURATION_MS);
+  }
+
+  dismissMessage(): void {
+    if (this.autoMessageDismissalTimeout) {
+      clearTimeout(this.autoMessageDismissalTimeout);
+    }
+    this.messageModalIsShown = false;
+    this.messageModalIsDismissed = true;
+    this.checkpointNodesAreVisible = false;
+    setTimeout(() => {
+      this.messageModalIsDismissed = false;
+    }, MESSAGE_MODAL_APPROX_TRIGGER_AND_DISMISSAL_DURATION_MS);
+  }
+
+  dismissMiniMessage(): void {
+    this.miniMessageTooltipIsShown = false;
+    this.miniMessageTooltipIsDismissed = true;
+    setTimeout(() => {
+      this.miniMessageTooltipIsDismissed = false;
+    }, MESSAGE_MODAL_APPROX_TRIGGER_AND_DISMISSAL_DURATION_MS);
+  }
+
+  resetTimer(): void {
+    if (!this.shouldDisplayFullScaleMessage) {
+      return;
+    }
+    this.checkpointTimer = document.querySelector(
+      '.checkpoint-celebration-modal-timer') as SVGPolylineElement;
```

Comment:
Ideally, we should not be using a document wide query like this in a component. What would happen if we have multiple elements with the same class name or multiple checkpoint-celebration-modal components on the page? Can we use ngStyle or similar in the SVG element change the values directly here?

Reply:
I actually need to access the element in the component to be able to trigger a reflow. I've modified it so instead of fetching the element by its class, I'm now using a template reference variable. PTAL!